libpod/sqlite_state.go

```diff
@@ -0,0 +1,2535 @@
+	)
+
+	// Some fields may be empty, indicating they are set to the default.
+	// If so, grab the default from c/storage for them.
+	if runtimeGraphRoot == "" {
+		runtimeGraphRoot = storeOpts.GraphRoot
+	}
+	if runtimeRunRoot == "" {
+		runtimeRunRoot = storeOpts.RunRoot
+	}
+	if runtimeGraphDriver == "" {
+		runtimeGraphDriver = storeOpts.GraphDriverName
+	}
+
+	row := s.conn.QueryRow("SELECT Os, StaticDir, TmpDir, GraphRoot, RunRoot, GraphDriver, VolumeDir FROM DBConfig;")
+
+	if err := row.Scan(&os, &staticDir, &tmpDir, &graphRoot, &runRoot, &graphDriver, &volumePath); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// Need to create runtime config info in DB
+			tx, err := s.conn.Begin()
+			if err != nil {
+				return fmt.Errorf("beginning DB config transaction: %w", err)
+			}
+			defer func() {
+				if defErr != nil {
+					if err := tx.Rollback(); err != nil {
+						logrus.Errorf("Rolling back transaction to create DB config: %v", err)
+					}
+				}
+			}()
+
+			if _, err := tx.Exec(createRow, 1, schemaVersion, runtimeOS,
+				runtimeStaticDir, runtimeTmpDir, runtimeGraphRoot,
+				runtimeRunRoot, runtimeGraphDriver, runtimeVolumePath); err != nil {
+				return fmt.Errorf("adding DB config row: %w", err)
+			}
+
+			if err := tx.Commit(); err != nil {
+				return fmt.Errorf("committing DB config transaction: %w", err)
+			}
+
+			return nil
+		}
+
+		return fmt.Errorf("retrieving DB config: %w", err)
+	}
+
+	checkField := func(fieldName, dbVal, ourVal string) error {
+		if dbVal != ourVal {
+			return fmt.Errorf("database %s %q does not match our %s %q: %w", fieldName, dbVal, fieldName, ourVal, define.ErrDBBadConfig)
+		}
+
+		return nil
+	}
+
+	if err := checkField("os", os, runtimeOS); err != nil {
+		return err
+	}
+	if err := checkField("static dir", staticDir, runtimeStaticDir); err != nil {
+		return err
+	}
+	if err := checkField("tmp dir", tmpDir, runtimeTmpDir); err != nil {
+		return err
+	}
+	if err := checkField("graph root", graphRoot, runtimeGraphRoot); err != nil {
+		return err
+	}
+	if err := checkField("run root", runRoot, runtimeRunRoot); err != nil {
+		return err
+	}
+	if err := checkField("graph driver", graphDriver, runtimeGraphDriver); err != nil {
+		return err
+	}
+	if err := checkField("volume path", volumePath, runtimeVolumePath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetContainerName returns the name of the container associated with a given
+// ID. Returns ErrNoSuchCtr if the ID does not exist.
+func (s *SQLiteState) GetContainerName(id string) (string, error) {
+	if id == "" {
+		return "", define.ErrEmptyID
+	}
+
+	if !s.valid {
+		return "", define.ErrDBClosed
+	}
+
+	var name string
+
+	row := s.conn.QueryRow("SELECT Name FROM ContainerConfig WHERE ID=?;", id)
+	if err := row.Scan(&name); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return "", define.ErrNoSuchCtr
+		}
+
+		return "", fmt.Errorf("looking up container %s name: %w", id, err)
+	}
+
+	return name, nil
```

Comment:
would it be faster to put the return name after the query and deal with the errors afterwards?

Reply:
Can you describe this better? I can't really visualize what you're looking for.